gpt_index/tools/migrate_v1_to_v2.py

```diff
@@ -0,0 +1,213 @@
+
+    struct_v2 = V2IndexDict(nodes_dict=nodes_dict_v2, doc_id_dict=doc_id_dict_v2, embeddings_dict=struct.embeddings_dict)
+    nodes_v2 = [node_to_v2(node) for node in struct.nodes_dict.values()]
+    return struct_v2, nodes_v2
+
+def kg_to_v2(struct: KG) -> Tuple[V2KG, List[V2Node]]:
+    struct_v2 = V2KG(
+        table=struct.table,
+        rel_map=struct.rel_map,
+        embedding_dict=struct.embedding_dict,
+    )
+    nodes_v2 = [node_to_v2(node) for node in struct.text_chunks.values()]
+    return struct_v2, nodes_v2
+
+def convert_to_v2(index_struct: IndexStruct, docstore: DocumentStore) -> Tuple[V2IndexStruct, DocumentStore]:
+    if isinstance(index_struct, IndexGraph):
+        struct_v2, nodes_v2 = index_graph_to_v2(index_struct)
+    elif isinstance(index_struct, IndexList):
+        struct_v2, nodes_v2 = index_list_to_v2(index_struct)
+    elif isinstance(index_struct, IndexDict):
+        struct_v2, nodes_v2 = index_dict_to_v2(index_struct)
+    elif isinstance(index_struct, KG):
+        struct_v2, nodes_v2 = kg_to_v2(index_struct)
+    else:
+        raise NotImplementedError(f"Cannot migrate {type(index_struct)} yet.")
+    
+    docstore.add_documents(nodes_v2)
+    return struct_v2, docstore
+    
+
+def load_v1_index_struct_in_docstore(file_dict):
+    index_struct_id = file_dict[INDEX_STRUCT_ID_KEY]
+    docstore = DocumentStore.load_from_dict(
+        file_dict[DOC_STORE_KEY],
+        type_to_struct=INDEX_STRUCT_TYPE_TO_INDEX_STRUCT_CLASS
+    )
+    index_struct = docstore.get_document(index_struct_id)
+    return index_struct, docstore
+
+
+def load_v1_index_struct_separate(file_dict, index_struct_type: IndexStructType):
+    index_struct_cls = INDEX_STRUCT_TYPE_TO_INDEX_STRUCT_CLASS[index_struct_type]
+    index_struct = index_struct_cls.from_dict(file_dict[INDEX_STRUCT_KEY])
+    docstore = DocumentStore.load_from_dict(
+        file_dict[DOC_STORE_KEY],
+        type_to_struct=INDEX_STRUCT_TYPE_TO_INDEX_STRUCT_CLASS
+    )
+    return index_struct, docstore
+
+
+def load_v1(file_dict: dict, index_struct_type: Optional[IndexStructType] = None) -> Tuple[IndexStruct, DocumentStore]:
+    if INDEX_STRUCT_KEY in file_dict:
+        assert index_struct_type is not None, 'Must specify index_struct_type to load.'
+        index_struct, docstore = load_v1_index_struct_separate(file_dict, index_struct_type)
+    elif INDEX_STRUCT_ID_KEY in file_dict:
+        index_struct, docstore = load_v1_index_struct_in_docstore(file_dict)
+    else:
+        raise ValueError("index_struct or index_struct_id must be provided.")
+    return index_struct, docstore
+
+
+def save_v2(index_struct: V2IndexStruct, docstore: DocumentStore) -> dict:
+    return {
+        INDEX_STRUCT_KEY: index_struct.to_dict(),
+        DOC_STORE_KEY: docstore.serialize_to_dict(),
+    }
+
+
+def main(in_path: str, index_struct_type: Optional[IndexStructType]=None, out_path: Optional[str] = None, encoding: str = 'ascii'):
```

Comment:
maybe give a line or two of example usage?

Reply:
also tbh i don't think running via command line is best UX for migration

ideally we support importing functions from this file for use in a jupyter notebook. which probably shouldn't require too many changes except renaming `main` to something more descriptive

Reply:
renamed